routers/web/web.go

```diff
@@ -1321,18 +1321,22 @@ func registerWebRoutes(m *web.Router) {
 					Post(web.Bind(forms.EditRepoFileForm{}), repo.NewFilePost)
 				m.Combo("/_delete/*").Get(repo.DeleteFile).
 					Post(web.Bind(forms.DeleteRepoFileForm{}), repo.DeleteFilePost)
-				m.Combo("/_upload/*", repo.MustBeAbleToUpload).Get(repo.UploadFile).
+				m.Combo("/_upload/*", context.MustBeAbleToUpload()).Get(repo.UploadFile).
 					Post(web.Bind(forms.UploadRepoFileForm{}), repo.UploadFilePost)
 				m.Combo("/_diffpatch/*").Get(repo.NewDiffPatch).
 					Post(web.Bind(forms.EditRepoFileForm{}), repo.NewDiffPatchPost)
+				m.Combo("/_fork_to_edit/*").
+					Post(web.Bind(forms.ForkToEditRepoFileForm{}), repo.ForkToEditFilePost)
+			}, context.MustEnableEditor())
+			m.Group("", func() {
 				m.Combo("/_cherrypick/{sha:([a-f0-9]{7,64})}/*").Get(repo.CherryPick).
 					Post(web.Bind(forms.CherryPickForm{}), repo.CherryPickPost)
-			}, context.RepoRefByType(git.RefTypeBranch), context.CanWriteToBranch(), repo.WebGitOperationCommonData)
-			m.Group("", func() {
-				m.Post("/upload-file", repo.UploadFileToServer)
-				m.Post("/upload-remove", web.Bind(forms.RemoveUploadFileForm{}), repo.RemoveUploadFileFromServer)
-			}, repo.MustBeAbleToUpload, reqRepoCodeWriter)
```

Comment:
The `reqRepoCodeWriter` permission check is missing? Then any code reader could call these route handlers?

Reply:
Sorry about that, hopefully it's fixed now.

Reply:
The permission check should be like `reqRepoCodeWriter || (reqRepoCodeReader && reqRepoPullRequestReader)`

Reply:
Maybe you can move the permission checking here.

Reply:
I'm not sure what you mean? The permission check is now in `MustBeAbleToUpload`.

`reqRepoCodeWriter` and `reqRepoCodeReader` are functions, and I don't think I can combine them with `||` and `&&` operators.

The check I added in `MustBeAbleToUpload` is also different. It requires having a repository that can be written to, which I think matches what it was before. With `reqRepoCodeReader && reqRepoPullRequestReader` a user would be able to upload files, even if they have no way to create their own repository.